Bottom: c83278bbf22d3a809e54997ec81d3c56f54a3395
Top:    5b8ec8160f6bf6c8e4811e199d91c8e369fa98e3
Author: Roland Stigge <stigge@antcom.de>
Date:   2012-03-29 23:31:16 +0200

Fix i2c transfers by reordering i2c start and irq

This patch fixed i2c transfers by enabling interrupts only after i2c_pnx_start

Signed-off-by: Roland Stigge <stigge@antcom.de>


---

diff --git a/drivers/i2c/busses/i2c-pnx.c b/drivers/i2c/busses/i2c-pnx.c
index 585971a..e9d66b6 100644
--- a/drivers/i2c/busses/i2c-pnx.c
+++ b/drivers/i2c/busses/i2c-pnx.c
@@ -528,16 +528,16 @@ i2c_pnx_xfer(struct i2c_adapter *adap, struct i2c_msg *msgs, int num)
 		/* initialize the completion var */
 		init_completion(&alg_data->mif.complete);
 
-		/* Enable master interrupt */
-		iowrite32(ioread32(I2C_REG_CTL(alg_data)) | mcntrl_afie |
-				mcntrl_naie | mcntrl_drmie,
-			  I2C_REG_CTL(alg_data));
-
 		/* Put start-code and slave-address on the bus. */
 		rc = i2c_pnx_start(addr, alg_data);
 		if (rc < 0)
 			break;
 
+		/* Enable master interrupt */
+		iowrite32(ioread32(I2C_REG_CTL(alg_data)) | mcntrl_afie |
+				mcntrl_naie | mcntrl_drmie,
+			  I2C_REG_CTL(alg_data));
+
 		/* Wait for completion */
 		wait_for_completion(&alg_data->mif.complete);
